### ISSUE TITLE
Look for other signals like EXIT, to exit the script

### ISSUE THREAD

**Issue by reportaman** [gpakosz/.tmux#509]
Following up on your comment https://github.com/tmux/tmux/issues/2778#issuecomment-880229283, yes that will be great! I still see the scripts keep running for some time after exiting tmux, and the original problem mentioned there still exists.

**Comment by gpakosz** [comment]
Hello @reportaman 👋

Trapping `EXIT` doesn't work. This is why I didn't apply this change already.

Contrary to what you say, the process do exit after `60s` after tmux has exited.

**Comment by reportaman** [comment]
@gpakosz You can also detect if tmux server exited using:

1) On macOS with GNU coreutils installed (``brew install coreutils``):
```
gtail --pid=$(tmux display-message -pF '#{pid}') -f /dev/null && echo "server exited"
```
2) On linux:
```
tail --pid=$(tmux display-message -pF '#{pid}') -f /dev/null && echo "server exited"
```

I have a feature request/issue on tmux related to detecting server exit: https://github.com/tmux/tmux/issues/2838 to perhaps add ``hook_pid`` that would ensure that the pid of correct tmux server is returned. Not sure if multiple hooks are allowed or appending to users hook will be allowed.

Yep I confirm I haven't seen the processes running more than 60 seconds recently...

**Comment by reportaman** [comment]
Also in my testing, ``gtail --pid``/``tail --pid`` does not consume any noticeable CPU performance. All the while CPU usage remains at ``0.0`` when observed in ``htop``.

**Comment by gpakosz** [comment]
It's just simpler to let the `sleep` process terminate gracefully, which happens at most 60s after tmux has exited 🤷‍♂️

**Comment by reportaman** [comment]
@gpakosz Hmm, I do like to use ``gpakosz/.tmux`` even while doing development for in tmux -- in the process I kill tmux and restart it immediately to see if things work well. So yeah for average tmux use 60s delay doesn't matter I guess, but if one is trying to develop plugin for it, then each time they will have to exit tmux and use ``kill -s TRAP <pid_of_.tmux_sh_sctipt>``... :)

**Comment by gpakosz** [comment]
Even if you're developing tmux plugins, these `sleep` process waiting to exit don't do any arm and consume 0% CPU 🤷‍♂️

**Comment by reportaman** [comment]
@gpakosz Not sure if you/@nicm understood my main concern in previous question. Main issue is unless I kill your sleeping scripts manually, a ``tmux kill-server`` seems to close one server, but reopen another with same name, as if there were a duplication of sorts. Why are there two ``/private/tmp/tmux-501/default`` lines in https://github.com/tmux/tmux/issues/2778#issue-944772844?

> The behavior of >>tmux kill-server also changes. If I had opened alacritty only once, running >>tmux kill-server kills server and closes the window. But after two /private/tmp/tmux-501/default entries appear, running >>tmux kill-server seems to kill one of the /private/tmp/tmux-501/default, and tmux session (with same name) continues to exist. All the while, there seems to be only one session visible in session list loaded by pressing <tmux_prefix>s.

**Comment by gpakosz** [comment]
Does the problem you're mentioning with Alacritty doesn't happen when not using Oh my tmux!?

**Comment by gpakosz** [comment]
Hello @reportaman 👋

You can try the `gh-509` branch

**Comment by reportaman** [comment]
@gpakosz Yes, it worked well! Thanks. Btw, in both old code & new code if the user issues ``kill <tmuxpid>`` instead of ``tmux kill-server``, there is one sleeping ohmytmux process that continues to exist beyond 60s.

In old code it is:
```
sh -c trap 'exit 0' TERM; while :; do nice cut -c3- ~/.tmux.conf | sh -s _uptime; sleep 60; done
```

In new code it is:
```
sh -c trap '[ -n "$sleep_pid" ] && kill -9 $sleep_pid; exit 0' TERM; while :; do nice cut -c3- ~/.tmux.conf | sh -s _uptime; sleep 60 & sleep_pid=$!; wait $sleep_pid; sleep_pid= ; done
```

**Comment by reportaman** [comment]
Oh happens if user kills tmux server with `` kill -s TRAP <tmux_server_pid>``.

**Comment by gpakosz** [comment]
It doesn't make sense to send `SIGTRAP` to a process with the `kill` command 🤷‍♂️

**Comment by reportaman** [comment]
@gpakosz Ok, got it... I started using after seeing some unruly command line utility not exiting with default kill signal. Your changes are good to otherwise...

**Comment by reportaman** [comment]
Not sure if it can be caught but https://github.com/tmux/tmux/issues/2917 causes the ohmytmux process to keep running beyond 1 minute. IMO, all of the ugly cases will get covered if instead of catching certain signals, ohmytmux looks at the server pid, and uses tail like here: https://github.com/gpakosz/.tmux/issues/509#issuecomment-903045116 to know when the server pid no longer exists. I do that in some of my local tmux code, and it works well.

**Comment by gpakosz** [comment]
> IMO, all of the ugly cases will get covered if instead of catching certain signals, ohmytmux looks at the server pid, and uses tail like here

The `sleep` processes exiting at most 60s after the tmux process has exited cause no problem at all.

The changes I pushed in the `gh-509` branch make these `sleep` process exit instantly when exiting tmux or when using `kill-server` but complicate things for no strong appealing reason.

I don't know what "all the ugly cases" you're mentioning are. As I sad, sending `SIGTRAP` manually to a pid doesn't make sense so I don't know what the other ugly cases are. In any case, I don't want Oh my tmux! to have a dependency on `tail` or `gtail` on macOS.

Finally, about tmux/tmux#2917 well if really tmux `HEAD` crashes when doing `tmux splitw -dI`, well this won't be a problem once tmux itself is fixed.

**Comment by reportaman** [comment]
@gpakosz Wow, thats what I was expecting, but is not the case. In my ohmytmux local repo, I am on branch ``gh-509``. There is no ohmytmux process like one you mention when I do ``ps aux |grep sleep``. I am on macOS 11.6, aarch64. Am surprised...

**Comment by reportaman** [comment]
@gpakosz Not sure what you are asking me to do. Give it another try?

**Comment by reportaman** [comment]
@gpakosz "When tmux restarts, Oh my tmux kills the sleep process anyway." What happens if user has multiple different tmux servers that are started and closed over time (with overlapping lifetimes)? Does ohmytmux of one server interfere with another?

**Comment by gpakosz** [comment]
For now Oh my tmux supports a single server per machine, which I believe corresponds to 99.99% of use cases

**Comment by reportaman** [comment]
@gpakosz Ok sure... For work use, 99.99% would be a bit high. Not sure if @nicm would guess.
